TLS version incompatibility when using latest JDKs with MySQL <8

Hi! I've been working on fixing [an issue in Testcontainers' test suite](https://github.com/testcontainers/testcontainers-java/issues/4058), and found a related issue here as well.

https://bugs.openjdk.java.net/browse/JDK-8202343 has disabled TLS1 and TLS1.1 in the latest versions of JDKs, and this seems to be particularly troublesome for connections to legacy versions of MySQL (e.g. 5.6, 5.7).

Following this (quite surprising) change to the JDK, the `SslBridgeHandler` code [here](https://github.com/mirromutth/r2dbc-mysql/blob/main/src/main/java/dev/miku/r2dbc/mysql/client/SslBridgeHandler.java#L195-L212) is now capable of producing an invalid set of TLS protocols when using:

* An up to date OpenJDK, e.g. `1.11.0-11`
* A MySQL server version <=8.0.4 (even though, per [docs](https://dev.mysql.com/doc/connector-j/8.0/en/connector-j-reference-using-ssl.html) and my own testing, >=5.7.28 and >=5.6.46 are capable of using TLS1.2)

With this combination of new and old, `r2dbc-mysql` is choosing to use TLS1.1/TLS1, which quickly fails with:

```
Caused by: javax.net.ssl.SSLHandshakeException: No appropriate protocol (protocol is disabled or cipher suites are inappropriate)
```

I'd imagine that a small tweak to the `isEnabledTls1_2` function would allow TLS1.2 to be used with older MySQL versions that can use it, so that the library is less inclined to try to use one of the disabled protocols. It might be good to fail-fast as well, with an informative exception when this incompatibility arises.

--- a/src/main/java/dev/miku/r2dbc/mysql/MySqlConnectionConfiguration.java
+++ b/src/main/java/dev/miku/r2dbc/mysql/MySqlConnectionConfiguration.java
@@ -433,11 +433,13 @@ public final class MySqlConnectionConfiguration {
         }
 
         /**
-         * Configure the socket timeout.  Default no timeout.
+         * Configure the socket timeout, only for compatibility with {@code socketTimeout} property in the
+         * JDBC driver.  In fact, {@code SO_TIMEOUT} has effect only for OIO socket transport.  Default no
+         * timeout.
          *
          * @param socketTimeout the socket timeout, or {@code null} if has no timeout.
          * @return this {@link Builder}.
-         * @since 0.8.3
+         * @since 0.8.6
          */
         public Builder socketTimeout(@Nullable Duration socketTimeout) {
             this.socketTimeout = socketTimeout;
@@ -782,6 +784,6 @@ public final class MySqlConnectionConfiguration {
             return sslMode;
         }
 
-        private Builder() {}
+        private Builder() { }
     }
 }

--- a/src/main/java/dev/miku/r2dbc/mysql/ServerVersion.java
+++ b/src/main/java/dev/miku/r2dbc/mysql/ServerVersion.java
@@ -24,7 +24,11 @@ import static dev.miku.r2dbc.mysql.util.AssertUtils.requireNonNull;
  */
 public final class ServerVersion implements Comparable<ServerVersion> {
 
-    private static final String[] ENTERPRISES = new String[] { "enterprise", "commercial", "advanced" };
+    private static final String ENTERPRISE = "enterprise";
+
+    private static final String COMMERCIAL = "commercial";
+
+    private static final String ADVANCED = "advanced";
 
     /**
      * Unresolved/origin version pattern, do NOT use it on {@link #hashCode()}, {@link #equals(Object)} or
@@ -95,18 +99,14 @@ public final class ServerVersion implements Comparable<ServerVersion> {
 
     /**
      * Checks if the version is enterprise edition.
+     * <p>
+     * Notice: it is unstable API, should not be used outer than {@literal r2dbc-mysql}.
      *
      * @return if is enterprise edition.
      */
     public boolean isEnterprise() {
-        for (String enterprise : ENTERPRISES) {
-            // Maybe should ignore case?
-            if (origin.contains(enterprise)) {
-                return true;
-            }
-        }
-
-        return false;
+        // Maybe should ignore case?
+        return origin.contains(ENTERPRISE) || origin.contains(COMMERCIAL) || origin.contains(ADVANCED);
     }
 
     @Override

--- a/src/main/java/dev/miku/r2dbc/mysql/client/SslBridgeHandler.java
+++ b/src/main/java/dev/miku/r2dbc/mysql/client/SslBridgeHandler.java
@@ -51,15 +51,19 @@ final class SslBridgeHandler extends ChannelDuplexHandler {
 
     private static final Logger logger = LoggerFactory.getLogger(SslBridgeHandler.class);
 
-    /**
-     * Lowest version of community edition for TLSv1.2 support.
-     */
-    private static final ServerVersion TLS1_2_COMMUNITY_VER = ServerVersion.create(8, 0, 4);
+    private static final String[] TLS_PROTOCOLS = new String[] {
+        TlsVersions.TLS1_3, TlsVersions.TLS1_2, TlsVersions.TLS1_1, TlsVersions.TLS1
+    };
 
-    /**
-     * Lowest version of enterprise edition for TLSv1.2 support. See {@link ServerVersion#isEnterprise()}.
-     */
-    private static final ServerVersion TLS1_2_ENTERPRISE_VER = ServerVersion.create(5, 6, 0);
+    private static final String[] OLD_TLS_PROTOCOLS = new String[] { TlsVersions.TLS1_1, TlsVersions.TLS1 };
+
+    private static final ServerVersion VER_5_6_0 = ServerVersion.create(5, 6, 0);
+
+    private static final ServerVersion VER_5_6_46 = ServerVersion.create(5, 6, 46);
+
+    private static final ServerVersion VER_5_7_0 = ServerVersion.create(5, 7, 0);
+
+    private static final ServerVersion VER_5_7_28 = ServerVersion.create(5, 7, 28);
 
     private final ConnectionContext context;
 
@@ -116,7 +120,7 @@ final class SslBridgeHandler extends ChannelDuplexHandler {
                     "' could not be verified"));
                 return;
             }
-            // Otherwise verify success, continue subsequence logic.
+            // Otherwise, verify success, continue subsequence logic.
         }
 
         if (mode != SslMode.TUNNEL) {
@@ -147,7 +151,7 @@ final class SslBridgeHandler extends ChannelDuplexHandler {
                 ctx.pipeline().remove(NAME);
                 break;
         }
-        // Ignore another custom SSL states because they are useless.
+        // Ignore another unknown SSL states because it should not throw an exception.
     }
 
     private HostnameVerifier hostnameVerifier() {
@@ -163,7 +167,7 @@ final class SslBridgeHandler extends ChannelDuplexHandler {
     }
 
     private static SslContextBuilder buildContext(MySqlSslConfiguration ssl, ServerVersion version) {
-        SslContextBuilder builder = withTlsVersion(SslContextBuilder.forClient(), ssl, version);
+        SslContextBuilder builder = withTlsVersion(ssl, version);
         String sslKey = ssl.getSslKey();
 
         if (sslKey != null) {
@@ -184,7 +188,7 @@ final class SslBridgeHandler extends ChannelDuplexHandler {
             if (sslCa != null) {
                 builder.trustManager(new File(sslCa));
             }
-            // Otherwise use default algorithm with trust manager.
+            // Otherwise, use default algorithm with trust manager.
         } else {
             builder.trustManager(InsecureTrustManagerFactory.INSTANCE);
         }
@@ -192,23 +196,29 @@ final class SslBridgeHandler extends ChannelDuplexHandler {
         return ssl.customizeSslContext(builder);
     }
 
-    private static SslContextBuilder withTlsVersion(SslContextBuilder builder, MySqlSslConfiguration ssl,
-        ServerVersion version) {
+    private static SslContextBuilder withTlsVersion(MySqlSslConfiguration ssl, ServerVersion version) {
+        SslContextBuilder builder = SslContextBuilder.forClient();
         String[] tlsProtocols = ssl.getTlsVersion();
 
         if (tlsProtocols.length > 0) {
             builder.protocols(tlsProtocols);
-        } else if (isEnabledTls1_2(version)) {
-            builder.protocols(TlsVersions.TLS1_2, TlsVersions.TLS1_1, TlsVersions.TLS1);
+        } else if (isCurrentTlsEnabled(version)) {
+            builder.protocols(TLS_PROTOCOLS);
         } else {
-            builder.protocols(TlsVersions.TLS1_1, TlsVersions.TLS1);
+            // Not sure if we need to check the JDK version, suggest not.
+            logger.warn("MySQL {} does not support TLS1.2, and TLS1.1 is disabled in latest JDKs", version);
+            builder.protocols(OLD_TLS_PROTOCOLS);
         }
 
         return builder;
     }
 
-    private static boolean isEnabledTls1_2(ServerVersion version) {
-        return version.isGreaterThanOrEqualTo(TLS1_2_COMMUNITY_VER) ||
-            (version.isGreaterThanOrEqualTo(TLS1_2_ENTERPRISE_VER) && version.isEnterprise());
+    private static boolean isCurrentTlsEnabled(ServerVersion version) {
+        // See also https://dev.mysql.com/doc/connector-j/8.0/en/connector-j-reference-using-ssl.html
+        // Quoting fragment: TLSv1,TLSv1.1,TLSv1.2,TLSv1.3 for MySQL Community Servers 8.0, 5.7.28 and
+        // later, and 5.6.46 and later, and for all commercial versions of MySQL Servers.
+        return version.isGreaterThanOrEqualTo(VER_5_7_28)
+            || (version.isGreaterThanOrEqualTo(VER_5_6_46) && version.isLessThan(VER_5_7_0))
+            || (version.isGreaterThanOrEqualTo(VER_5_6_0) && version.isEnterprise());
     }
 }

--- a/src/main/java/dev/miku/r2dbc/mysql/constant/TlsVersions.java
+++ b/src/main/java/dev/miku/r2dbc/mysql/constant/TlsVersions.java
@@ -23,28 +23,27 @@ public final class TlsVersions {
 
     /**
      * TLS version 1.0, not recommended. All available MySQL versions support it.
+     * <p>
+     * Notice: it has been disabled in the latest version of JDKs.
      */
     public static final String TLS1 = "TLSv1";
 
     /**
      * TLS version 1.1. All available MySQL versions support it.
+     * <p>
+     * Notice: it has been disabled in the latest version of JDKs.
      */
     public static final String TLS1_1 = "TLSv1.1";
 
     /**
-     * TLS version 1.2. It is supported in 8.0.4 (or higher) Community Edition, or 5.6.0 (or higher)
-     * Enterprise Edition.
-     * <p>
-     * Note: The Enterprise Edition version will looks like {@literal 5.7.26-enterprise}.
+     * TLS version 1.2. Supported in MySQL community edition 8.0, 5.7.28 and later, and 5.6.46 and later, and
+     * for all enterprise editions of MySQL Servers.
      */
     public static final String TLS1_2 = "TLSv1.2";
 
     /**
-     * TLS version 1.3. It is available as of MySQL 8.0.16 or higher, and requires compiling MySQL using
-     * OpenSSL 1.1.1 or higher.
-     * <p>
-     * It will not be used unless user has set it up, because the driver cannot know which compiling library
-     * the MySQL server uses.
+     * TLS version 1.3. Supported in MySQL community edition 8.0, 5.7.28 and later, and 5.6.46 and later, and
+     * for all enterprise editions of MySQL Servers.
      */
     public static final String TLS1_3 = "TLSv1.3";
 

--- a/src/test/java/dev/miku/r2dbc/mysql/IntegrationTestSupport.java
+++ b/src/test/java/dev/miku/r2dbc/mysql/IntegrationTestSupport.java
@@ -78,7 +78,6 @@ abstract class IntegrationTestSupport {
         MySqlConnectionConfiguration.Builder builder = MySqlConnectionConfiguration.builder()
             .host("127.0.0.1")
             .connectTimeout(Duration.ofSeconds(3))
-            .socketTimeout(Duration.ofSeconds(3))
             .user("root")
             .password(password)
             .database("r2dbc")

